include/openenclave/bits/sgx/seal.h

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#ifndef _OE_BITS_SGX_SEAL_H
+#define _OE_BITS_SGX_SEAL_H
+
+#include <openenclave/bits/sgx/sgxtypes.h>
+
+OE_EXTERNC_BEGIN
+
+/**
+ * SGX specific seal settings
+ */
+typedef enum _oe_seal_setting_name_sgx
+{
+    OE_SEAL_SETTING_SGX_KEYNAME = 0x100,
+    OE_SEAL_SETTING_SGX_ISVSVN,
+    OE_SEAL_SETTING_SGX_CET_ATTRIBUTES_MASK,
+    OE_SEAL_SETTING_SGX_CPUSVN,
+    OE_SEAL_SETTING_SGX_ATTREMASK,
+    OE_SEAL_SETTING_SGX_XFRMMASK,
+    OE_SEAL_SETTING_SGX_MISCMASK,
+    OE_SEAL_SETTING_SGX_CONFIGSVN,
+    OE_SEAL_SETTING_SGX_MAX
+} oe_seal_setting_name_sgx_t;
+
+#define _DEFINE_SEAL_SETTING_SGX(name, type)   \
+    typedef struct _oe_seal_setting_sgx_##name \
+    {                                          \
+        oe_seal_setting_header_t header;       \
+        type name;                             \
+    } oe_seal_setting_sgx_##name##_t
+
+_DEFINE_SEAL_SETTING_SGX(keyname, uint16_t);
+_DEFINE_SEAL_SETTING_SGX(isvsvn, uint16_t);
+_DEFINE_SEAL_SETTING_SGX(cet_attributes_mask, uint8_t);
+_DEFINE_SEAL_SETTING_SGX(attrmask, uint64_t);
+_DEFINE_SEAL_SETTING_SGX(xfrmmask, uint64_t);
+_DEFINE_SEAL_SETTING_SGX(miscmask, uint32_t);
+_DEFINE_SEAL_SETTING_SGX(configsvn, uint16_t);
+
+typedef struct _oe_seal_setting_sgx_cpusvn
+{
+    oe_seal_setting_header_t header;
+    uint8_t cpusvn[SGX_CPUSVN_SIZE];
+} oe_seal_setting_sgx_cpusvn_t;
+
+#undef _DEFINE_SEAL_SETTING_SGX
+
+typedef struct _oe_seal_plugin_definition_sgx
+{
+    const oe_uuid_t id;
+
+    oe_result_t (*seal_callback)(
+        sgx_key_request_t* keyrequest,
```

Comment:
The intention is to allow the top level `oe_seal()` to parse settings into `keyrequest` so that plug-ins don't have to duplicate that parsing code. Any settings not recognized by `oe_seal()` itself will be passed through to the plug-in. I agree that the top level should reserve a range for future extension. When it comes to plug-in specific settings, it's actually the pair of { `id`, `settings->name` }  (i.e., not `settings->name` alone) that uniquely identifies the setting, so I won't worry about collisions between plug-ins.

Reply:
`oe_seal()` should not maintain multiple TEE-specific contracts with different sealing implementations. This breaks dependency inversion and defeats the goal of a good extension model. 

Regarding preventing collisions at the setting ID level: one of the features of the proposed in the plugin-model strawman is that a developer can write their sealing code once, providing all options on all platforms, and have the extension model fall through to the appropriate plugin that can handle it, more like the attestation plugin model.

While we don't have to implement that feature now, assuming that sealing invocations must always be coupled to an explicit uuid target (where NULL is also a single default target with no fallthrough) can be a problem; it can make it hard to implement the feature in the future where we may already have an ecosystem of plugins that trample each other. 

We could work around this by forcing the namespacing into the setting itself (this is what my original suggestion for the extensibility model does by having plugins declare their own uuid-qualified custom data structure, as a separate set of inputs from the OE-defined seal settings) and along those lines, it may be an option to just add an optional uuid member to the basic `oe_seal_setting_header_t` itself as the need arises in the future. 

@dthaler any thoughts on API design?